runtime/fiber.c

```diff
@@ -1117,3 +1133,318 @@ CAMLexport void caml_raise_unhandled_effect (value effect)
+    res->base.mask = parent->base.mask;
+    res->base.count = parent->base.count;
+    res->base.bindings = new_bindings;
+  }
+  return res;
+fail_alloc_bindings:
+  caml_stat_free(res);
+fail_alloc:
+  return NULL;
+}
+
+CAMLexport void caml_dynamic_delete_thread(dynamic_thread_t thread)
+{
+  if (thread->base.bindings) {
+    caml_stat_free(thread->base.bindings);
+  }
+  caml_stat_free(thread);
+}
+
+CAMLexport void caml_dynamic_flush_thread(dynamic_thread_t thread)
+{
+  dynamic_flush_cache(thread);
+}
+
+CAMLexport void caml_dynamic_enter_thread(dynamic_thread_t thread)
+{
+  Caml_state->dynamic_bindings = thread;
+}
+
+CAMLexport void caml_dynamic_scan_thread_roots(dynamic_thread_t thread,
+                                               scanning_action f,
+                                               scanning_action_flags fflags,
+                                               void *fdata)
+{
+  for (size_t i = 0; i < DYNAMIC_CACHE_SIZE; ++i) {
+    if (Is_bound(&thread->cache[i])) {
+      f(fdata, thread->cache[i].dyn, &thread->cache[i].dyn);
+      f(fdata, thread->cache[i].val, &thread->cache[i].val);
+    }
+  }
+  for (size_t i = 0; i < (size_t)thread->base.mask + 1; ++i) {
+    if (Is_bound(&thread->base.bindings[i])) {
+      f(fdata, thread->base.bindings[i].dyn, &thread->base.bindings[i].dyn);
+      f(fdata, thread->base.bindings[i].val, &thread->base.bindings[i].val);
+    }
+  }
+
+}
+
+CAMLprim value caml_dynamic_set_root(value dyn, value val)
+{
+  CAMLparam2(dyn, val); /* TODO: remove unless we can GC */
+  dynamic_thread_t thread = Caml_state->dynamic_bindings;
+  CAMLassert(thread);
+  bool res = dynamic_table_bind(&thread->base, dyn, val);
+  if (!res) {
+    caml_raise_out_of_memory();
+  }
+
+  /* invalidate cache */
+  uintnat hash = Hash_dyn(dyn);
+  uintnat index = hash & (DYNAMIC_CACHE_SIZE - 1);
+  binding_t entry = thread->cache + index;
+  if (entry->dyn == dyn) {
+    entry->dyn = Val_null;
+  }
+
+  CAMLreturn(Val_unit);
+}
+
+/* Get the current value of a dynamic variable. Does not allocate.
+ *
+ * Comments indicate the memory accesses on the fast path (loads from
+ * three cache lines; load #3 is dependent on load #2). */
+
+CAMLprim value caml_dynamic_get(value dyn)
+{
+  uintnat hash = Hash_dyn(dyn); /* load #1 */
+  uintnat index = hash & (DYNAMIC_CACHE_SIZE - 1);
+  dynamic_thread_t thread = Caml_state->dynamic_bindings; /* load #2 */
+  CAMLassert(thread);
+  binding_t entry = thread->cache + index; /* just arithmetic */
+  if (entry->dyn == dyn) { /* load #3; dependent on #2 */
+    return entry->val; /* load #4, should hit same cache line as #3 */
+  }
+  value val;
+  /* Not in cache; let's look at the fiber */
+  CAMLassert(Caml_state->current_stack);
+  struct stack_info *stack = Caml_state->current_stack;
+  while (stack) {
+    if (dyn == stack->dyn) {
+      val = stack->bind;
+      goto found;
+    }
+    stack = Stack_parent(stack);
```

Comment:
Probably doesn't matter, but could guard against finding a cycle

Reply:
I thought about putting Floyd's tortoise-and-hare algorithm in here, but decided against: if we ever get a cycle in the stacks graph then everything will get stuck in an infinite loop (e.g. the GC) and that should be easy to find and fix. If we do want to guard against it, we should probably do so in `caml_scan_stack` rather than here.